### PR TITLE
loadbalancer and loadbalancer claim proposal

### PR DESCRIPTION
**Pull request by mqliang**
@ddysher PTAL

### CODE REVIEW

File: docs/proposals/loadbalancer-proposal.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Loadbalancer and LoadbalancerClaim proposal
+----
+
+### Overview
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md:38]: `Overview` is not enough here, we need `background` kind of section; especially its relationship with existing ingress resource, and how it aligns with current kubernetes models (like computing for node/pod, storage for pv/pod, etc).

File: docs/proposals/loadbalancer-proposal.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+LoadBalancer is a commen resource provided by IaaS cloud provider, but currently
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: what's the problem with this ?  Better to spec it out

File: docs/proposals/loadbalancer-proposal.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+kuberntes does not expose cloud LoadBalancer resource to users explicitly, we
+just use it implicitly by "Loadbalancer" type Service. We proposed to:
+
+* explicitly expose cloud loadbalancer to users
+* introduce `LoadbalancerClaim` mechanism to allow user to create, use, and delete
+loadbalancer in a convienet way.
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: I think you should also mention ingress resource; that is, user creates ingress to actually consume loadbalancer

File: docs/proposals/loadbalancer-proposal.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+####  Goal
+* LoadBalancer and LoadBalancerClaim API
+* LoadBalancer provisioning
+* LoadBalancer scheduling based on bandwidth and iops request
+
+#### NoneGoal
+* LoadBalancerClass: different type(internet or intranet), different qos level etc
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: add space between char and (

File: docs/proposals/loadbalancer-proposal.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+* LoadBalancer scheduling over-commit
+
+
+### Design
+* Add a `LoadBalancer` API to expose LoadBalancer explicitly to user
+* Add bandwidth and iops as new resource type, to describe the resource capacity
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: should we leave room for other new types ?  And how do they fit into ur model?

File: docs/proposals/loadbalancer-proposal.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  of a LoadBalancer and describe the resource a LoadBalancerClaim request
+* When user create a Ingress resource, we need a LoadBalancer to satisfy the
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: Just describe LoadBalancerClaim and what's it for

File: docs/proposals/loadbalancer-proposal.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+  request, but which LoadBalancer to choose? we propose that Ingress "use" a
+  loadbalancer indirectly through LoadBalancerClaim
+* If use know exactly which LoadBalancer to use, he can create a LoadBalancerClaim
+  and bind it with the LoadBalancer manully. Otherwise, a loadbalancer-scheduler
+  will help user to make the decision based on the resource a LoadBalancerClaim
+  request and LoadBalancer's remaining capacity
+* LoadBalancerClaim->LoadBalancer binding is non-exclusive
+
+### Implement
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: Implementation

File: docs/proposals/loadbalancer-proposal.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+
+### Design
+* Add a `LoadBalancer` API to expose LoadBalancer explicitly to user
+* Add bandwidth and iops as new resource type, to describe the resource capacity
+  of a LoadBalancer and describe the resource a LoadBalancerClaim request
+* When user create a Ingress resource, we need a LoadBalancer to satisfy the
+  request, but which LoadBalancer to choose? we propose that Ingress "use" a
+  loadbalancer indirectly through LoadBalancerClaim
+* If use know exactly which LoadBalancer to use, he can create a LoadBalancerClaim
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: How do you manually bind? You should spell it out in ur design.

File: docs/proposals/loadbalancer-proposal.md (hunk 2 above)
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: s/commen/common

Comment (deyuanix) [reply docs/proposals/loadbalancer-proposal.md]: Not just IaaS cloud provider right? what about baremetal?

File: docs/proposals/loadbalancer-proposal.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Loadbalancer and LoadbalancerClaim proposal
+----
+
+### Overview
+
+LoadBalancer is a commen resource provided by IaaS cloud provider, but currently
+kuberntes does not expose cloud LoadBalancer resource to users explicitly, we
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: s/kuberntes/kubernetes

File: docs/proposals/loadbalancer-proposal.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+just use it implicitly by "Loadbalancer" type Service. We proposed to:
+
+* explicitly expose cloud loadbalancer to users
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: dont use `cloud` if you also want to cover baremetal

File: docs/proposals/loadbalancer-proposal.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+* introduce `LoadbalancerClaim` mechanism to allow user to create, use, and delete
+loadbalancer in a convienet way.
+
+####  Goal
+* LoadBalancer and LoadBalancerClaim API
+* LoadBalancer provisioning
```
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: auto provisioning or manual provisioning ?

File: docs/proposals/loadbalancer-proposal.md (hunk 6 above)
Comment (deyuanix) [inline docs/proposals/loadbalancer-proposal.md]: s/create/creates

**Comment by mqliang** [comment]
@ddysher Updated.PTAL.

**Comment by mqliang** [comment]
test

**Comment by deyuanix** [comment]
I'll merge then create a PR.

**Pull request merged**